design/proposal-controller-code-generation.md

```diff
@@ -0,0 +1,129 @@
+
+**Provider-specific discovery documents**: This mostly applies to [Google Discovery Documents]; Azure and Alibaba both directly support OpenAPI. If AWS has a service discovery format I haven’t been able to find it yet, but the other option here would be to use the [CloudFormation Resource Specification], particularly if we use CloudFormation for CRUD under the hood rather than individual AWS resource APIs.
+
+### Merge Patching / Injection / Overlays
+The generation of CRD types and reconciliation methods are challenging when the cloud provider interactions don’t map 1:1 with the CRD types. Instead of the kubebuilder paradigm of generating source code files that users modify, it would be interesting to explore injection points where some of the components outlined in “Output Targets” could be replaced or augmented, possibly in a middleware model. The code generation process should be modular and pluggable, allowing for customizations to the rendered code. Knowing exactly what form these overlays will take is going to require first prototyping the MRS->Go tools, but here are a few thoughts on what this could look like.
+
+It’s possible that we’ll wind up with injection points in the Frontend and Backend of the generation pipeline; manipulating CRD structure in the Frontend, and customizing code generation in the Backend. For instance in the Frontend, the json-schema object structure can be directly manipulated as in-memory go structs or json. Driving code generation by manipulating the MRS structure could be less fragile than directly working with the code generation routines in the backend. For a problem like customizing Reference resolution, we probably also need backend hooks where user-supplied code can be directly invoked.
+
+Alternatively, the generated code could follow a plugin registration pattern (keyed to a particular provider+resource+component) where developers can override/augment/append generated code without touching the generic code generation functions. Given that triggering code generation could have unintended side effects when provider metadata descriptions are revised, this could give developers more control to modify controllers without worrying about broader versioning concerns.
```

Comment:
Is this an either/or option with the injection points described above? It seems like both could be useful, but I may be misinterpreting the model here.

Reply:
Good point, "Alternatively" implies "either/or", but I meant this as an "and/or". My gut feeling is both types of hooks have their place but I need to get into the implementation more to know what design makes sense.